### DISCUSSION TITLE
`SSL_do_handshake` returns success on server side client certificate validation failure

### DISCUSSION THREAD

**Discussion by matthauck** [openssl/openssl#22229] [Q&A] [answered]
Migrating from openssl 1.0.2 (finally) to openssl 3, I discovered a difference in behavior on the client side of the handshake when the server fails to validate the client certificate. 

Here is a minimal example where a server requires a client cert but the client does not present one:

```c++
static int VerifyNone( int preVerifyOk, X509_STORE_CTX* ctx )
{
    return 1;
}
​
static void Run( X509* serverCert, EVP_PKEY* serverKey )
{
    SSL_CTX* serverCtx = SSL_CTX_new( TLS_server_method() );
    SSL_CTX* clientCtx = SSL_CTX_new( TLS_client_method() );
​
    SSL_CTX_set_verify( serverCtx, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, VerifyNone );
    SSL_CTX_set_verify( clientCtx, SSL_VERIFY_PEER, VerifyNone );
​
    BIO* ctos = BIO_new( BIO_s_mem() );
    BIO* stoc = BIO_new( BIO_s_mem() );
​
    BIO_set_mem_eof_return( ctos, -1 );
    BIO_set_mem_eof_return( stoc, -1 );
​
    SSL* client = SSL_new( clientCtx );
    SSL_set_connect_state( client );
    SSL_set_bio( client, ctos, stoc );
​
    SSL* server = SSL_new( serverCtx );
    SSL_set_accept_state( server );
    SSL_set_bio( server, stoc, ctos );
​
    SSL_use_PrivateKey( server, serverKey );
    SSL_use_certificate( server, serverCert );
​
    bool clientDone = false;
    bool serverDone = false;
​
    while ( !clientDone && !serverDone )
    {
        if ( !clientDone )
        {
            int status = SSL_do_handshake( client );
            std::cout << "client status: " << status << std::endl;
​
            if ( status == 1 )
            {
                std::cout << "client done" << std::endl;
                clientDone = true;
            }
            else
            {
                int error = SSL_get_error( client, status );
                if ( error != SSL_ERROR_WANT_READ && error != SSL_ERROR_WANT_WRITE )
                {
                    std::cout << "client error: " << error << " - " << ERR_reason_error_string( ERR_get_error() )
                              << std::endl;
                    clientDone = true;
                }
                else
                {
                    std::cout << "client wants read/write: " << error << std::endl;
                }
            }
        }
​
        if ( !serverDone )
        {
            int status = SSL_do_handshake( server );
            std::cout << "server status: " << status << std::endl;
​
            if ( status == 1 )
            {
                std::cout << "server done" << std::endl;
                serverDone = true;
            }
            else
            {
                int error = SSL_get_error( server, status );
                if ( error != SSL_ERROR_WANT_READ && error != SSL_ERROR_WANT_WRITE )
                {
                    std::cout << "server error: " << error << " - " << ERR_reason_error_string( ERR_get_error() )
                              << std::endl;
                    serverDone = true;
                }
                else
                {
                    std::cout << "server wants read/write: " << error << std::endl;
                }
            }
        }
    }
}
```

which yields the output:
```
client status: -1
client wants read/write: 2
server status: -1
server wants read/write: 2
client status: 1
client done
server status: -1
server error: 1 - peer did not return a certificate
```

The curious part is `client status: 1` -- the client thinks it is done after receiving the server hello. I would expect the client handshake to fail if the server does not validate the client certificate correctly.

- Is change a consequence of TLS 1.3? 
- Is there something else the client is supposed to be doing to make sure the server has accepted its client certificate before proceeding to try to send data? 
- Should it rely on a TLS alert callback to indicate the server rejected the client cert? 
- Or should the client just begin sending data and get an error on `SSL_write`?

Thanks!

**Comment by mattcaswell** [comment, marked as answer]
> Is change a consequence of TLS 1.3?

Yes. The handshake is considered complete at the point that an endpoint has both sent and received a "Finished" message. With TLSv1.3 the server sends its Finished message at the end of its first flight of messages in response to the ClientHello message. If a certificate is requested from the client, then the clients sends it back along with its own "Finished" message. At this point the client considers the handshake complete (it has both sent and received a Finished message) - but the server does not. If the server subsequently fails to verify the supplied certificate then it does not become visible to the client until after the handshake is complete (from the perspective of the client).

This asymmetry between when the client and server consider the handshake to be complete has always been present - but TLSv1.3 makes it visible in the case of client certificate verification failure (where it wasn't before).

> Is there something else the client is supposed to be doing to make sure the server has accepted its client certificate before proceeding to try to send data?
> Should it rely on a TLS alert callback to indicate the server rejected the client cert?
> Or should the client just begin sending data and get an error on SSL_write?

Yes the client should rely on an alert telling it that the server handshake failed. It seems reasonable to me that it should just get on with sending data on the assumption that the verification will succeed (presumably the client sent it in the expectation that verification would pass). That is unless the application protocol dictates some other behaviour.

**Reply by wolf-hunter404** [reply]
I have a similar situation.
I just want to test if the handshake is successful，do not send business data.
How to judge whether the handshake is really successful?

**Reply by mattcaswell** [reply]
You can only know for sure on the client side that the server considers the certificate acceptable, once you've started to receive data from the server. But whether the server will actually send that data is application protocol specific.

**Reply by wolf-hunter404** [reply]
That means, I can‘t use openSSL to determine whether the handshake is successful?

**Reply by mattcaswell** [reply]
No. OpenSSL does not know. This is by design of the TLSv1.3 protocol.

**Reply by wolf-hunter404** [reply]
I try to write data using SSL_write, but I'm not sure whether the server is HTTP Server or ordinary SSL Server.
It is also impossible to determine the format of the data to be sent.
When using SSL_set_msg_callback, ALERT only listens to close notify.
Is there any other way to confirm whether the handshake was actually successful?

**Reply by mattcaswell** [reply]
> Is there any other way to confirm whether the handshake was actually successful?

Like I said above the only way we know whether the server has accepted the certificate is when the server starts sending data back. Until then you can only assume that it has been accepted.
